### PR TITLE
[flang][cuda] Update target rewrite to work on gpu.func

### PR DESCRIPTION
**Pull request by clementval**
Update the pass so it can perform the signature rewrite on gpu.func.

### CODE REVIEW

**Review by jeanPerier — APPROVED** [review]
Small suggestion to avoid duplicating return op handling, LGTM otherwise.

File: flang/lib/Optimizer/CodeGen/TargetRewrite.cpp [hunk 1]
```diff
@@ -719,12 +726,15 @@ class TargetRewrite : public fir::impl::TargetRewritePassBase<TargetRewrite> {
...
+      convertSignature<mlir::func::ReturnOp>(fn);
     }
 
-    for (auto gpuMod : mod.getOps<mlir::gpu::GPUModuleOp>())
+    for (auto gpuMod : mod.getOps<mlir::gpu::GPUModuleOp>()) {
       for (auto fn : gpuMod.getOps<mlir::func::FuncOp>())
-        convertSignature(fn);
+        convertSignature<mlir::func::ReturnOp>(fn);
+      for (auto fn : gpuMod.getOps<mlir::gpu::GPUFuncOp>())
+        convertSignature<mlir::gpu::ReturnOp>(fn);
```
Comment (clementval) [inline flang/lib/Optimizer/CodeGen/TargetRewrite.cpp:736]:
```suggestion
        convertSignature<mlir::func::ReturnOp, mlir::func::FuncOp>(fn);
      for (auto fn : gpuMod.getOps<mlir::gpu::GPUFuncOp>())
        convertSignature<mlir::gpu::ReturnOp, mlir::gpu::GPUFuncOp>(fn);
```

File: flang/lib/Optimizer/CodeGen/TargetRewrite.cpp [hunk 2]
```diff
@@ -719,12 +726,15 @@ class TargetRewrite : public fir::impl::TargetRewritePassBase<TargetRewrite> {
       if (targetFeaturesAttr)
         fn->setAttr("target_features", targetFeaturesAttr);
 
-      convertSignature(fn);
+      convertSignature<mlir::func::ReturnOp>(fn);
```
Comment (clementval) [inline flang/lib/Optimizer/CodeGen/TargetRewrite.cpp:729]:
```suggestion
      convertSignature<mlir::func::ReturnOp, mlir::func::FuncOp>(fn);
```

File: flang/lib/Optimizer/CodeGen/TargetRewrite.cpp [hunk 3]
```diff
@@ -825,10 +838,10 @@ class TargetRewrite : public fir::impl::TargetRewritePassBase<TargetRewrite> {
                                                 rewriter->getUnitAttr()));
             newResTys.push_back(retTy);
           })
-          .Case<fir::RecordType>([&](fir::RecordType recTy) {
+          .template Case<fir::RecordType>([&](fir::RecordType recTy) {
             doStructReturn(func, recTy, newResTys, newInTyAndAttrs, fixups);
           })
-          .Default([&](mlir::Type ty) { newResTys.push_back(ty); });
+          .template Default([&](mlir::Type ty) { newResTys.push_back(ty); });
```
Comment (clementval) [inline flang/lib/Optimizer/CodeGen/TargetRewrite.cpp:844]:
```suggestion
          .Default([&](mlir::Type ty) { newResTys.push_back(ty); });
```

File: flang/lib/Optimizer/CodeGen/TargetRewrite.cpp [hunk 4]
```diff
@@ -898,13 +911,13 @@ class TargetRewrite : public fir::impl::TargetRewritePassBase<TargetRewrite> {
           .template Case<fir::RecordType>([&](fir::RecordType recTy) {
             doStructArg(func, recTy, newInTyAndAttrs, fixups);
           })
-          .Default([&](mlir::Type ty) {
+          .template Default([&](mlir::Type ty) {
```
Comment (clementval) [inline flang/lib/Optimizer/CodeGen/TargetRewrite.cpp:914]:
```suggestion
          .Default([&](mlir::Type ty) {
```

**Pull request merged**